Reduce memory usage in GpuFileFormatDataWriter and GpuDynamicPartitionDataConcurrentWriter

abellina opened the pull request:
Contributes to https://github.com/NVIDIA/spark-rapids/issues/6980

As part of https://github.com/NVIDIA/spark-rapids/issues/6980, we would like to make sure we don't let go of the semaphore with unspillable GPU memory. This PR is the first step towards that goal. It cleans up memory usage in some of the writers where we were holding onto the original table, and the splits, and then attempting to make things spillable as well in a retry block. It should help with OOM and spill situations while writing large data.

I did some `withResource` refactoring as well.

Draft because I am re-running tests to make sure I got all the leaks out.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -434,153 +418,129 @@ class GpuDynamicPartitionDataSingleWriter(
+            }
           }
           // write concat table
           if (!needSplitBatch(
```
nit: So...?? We might concat the tables together just to split them apart again?  It might be good to rename `needSplitBatch` to be a little more descriptive as to why we need to split a batch so it is clear what is happening.

abellina replied:
Will look, I was focused on reducing memory waste, didn't really go into the logic much yet.

abellina replied:
I renamed the function and tried to do some refactoring but I am finding it really not straight forward given the current design, and I think we should punt but _absolutely_ prioritize refactoring this.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -434,153 +418,129 @@ class GpuDynamicPartitionDataSingleWriter(
+        val savedStatus = updateCurrentWriterIfNeeded(partPath, cachesMap)
 
         if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
           // convert caches seq to tables and close caches seq
```
nit: Could we split this large function up into smaller functions?  It is quite hard to see what the else part of this if goes with.

There also appears to be a lot of common code between the various paths and it would be nice to make it common, or at least see what the differences are better. Like updating the currentWriters status and metrics/etc.

abellina replied:
sure thing

abellina commented:
I was chasing an existing leak: https://github.com/NVIDIA/spark-rapids/issues/8332. I will address review and then likely take on the known leak in a different PR.

abellina commented:
build

abellina commented:
build

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -434,179 +465,156 @@ class GpuDynamicPartitionDataSingleWriter(
-      if (cbKeys != null) {
-        cbKeys.close()
+      if (cachesMap.isDefined) {
+        savedStatus = cachesMap.get.get(currentPartPath)
+        if (savedStatus.isDefined) {
+          // first try to restore the saved writer status,
+          // `GpuDynamicPartitionDataConcurrentWriter` may already opened the writer, and may
+          // have pending caches
+          currentWriterStatus = savedStatus.get.writerStatus
+          // entire batch that is sorted, see a new partition, the old write status is useless
+          cachesMap.get.remove(previousPartPath)
```
```suggestion
        // entire batch that is sorted, see a new partition, the old write status is useless
        savedStatus = cachesMap.get.remove(currentPartPath)
        if (savedStatus.isDefined) {
          // first try to restore the saved writer status,
          // `GpuDynamicPartitionDataConcurrentWriter` may already opened the writer, and may
          // have pending caches
          currentWriterStatus = savedStatus.get.writerStatus
```

abellina replied:
oh wait, I can't do that. The code checks whether `currentPartPath` exists, but then removes `previousPartPath`.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -939,68 +954,65 @@ class GpuDynamicPartitionDataConcurrentWriter(
+
+    if (!shouldSplitToFitMaxRecordsPerFile(
+        maxRecordsPerFile, status.writerStatus.recordsInFile, batch.numRows())) {
+      statsTrackers.foreach(_.newBatch(status.writerStatus.outputWriter.path(), batch))
```
If any of these stat tracker callbacks throw, `batch` is leaked.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -939,68 +954,65 @@ class GpuDynamicPartitionDataConcurrentWriter(
+            status.writerStatus.recordsInFile = 0L
+          }
+          val cb = GpuColumnVector.from(b.getTable, dataTypes)
+          statsTrackers.foreach(_.newBatch(status.writerStatus.outputWriter.path(), cb))
```
If any of these stat tracker callbacks throw, `cb` is leaked.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -939,68 +954,65 @@ class GpuDynamicPartitionDataConcurrentWriter(
+            status.writerStatus.outputWriter = w
+            status.writerStatus.recordsInFile = 0L
+          }
+          val cb = GpuColumnVector.from(b.getTable, dataTypes)
```
Shouldn't we close `b` here and null out the corresponding `tabs` entry so we can close these early?  Looks like everything except the last entry in `tabs` is held on longer than it needs to be.

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -434,179 +465,156 @@ class GpuDynamicPartitionDataSingleWriter(
-        val partPath = combined._1
-
+    // We have an entire batch that is sorted, so we need to split it up by key
+    val (paths, splits) = withResource(cb) { _ =>
```
this is a nice catch, should be fixed. Thank you

jlowe replied:
This is still not fixed.  `cb` should be closed as soon as `getOutputColumnsAsTable` completes before we call `contiguousSplit` on that result.

abellina replied:
:( thanks again, you are right. I meant to do that, but I didn't do that. One sec

abellina replied:
Handled here: https://github.com/NVIDIA/spark-rapids/pull/8385/commits/f72f45ec0dee6340fd476416c3701ffff6581d13

abellina commented:
build

abellina commented:
CI failed due to the `repeatStaticCount` cuDF known issues.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -623,17 +635,23 @@ class GpuDynamicPartitionDataSingleWriter(
-          statsTrackers.foreach(_.newBatch(currentWriterStatus.outputWriter.path(), bc))
-          currentWriterStatus.recordsInFile += b.getRowCount()
+        splits(partIx) = null
+        withResource(part) { _ =>
+          writeTableUsingCurrentWriter(part.getTable, outDataTypes)
         }
```
This holds onto the table too long.  The writer internally will release resources as part of writing, but this holds onto the part table throughout the entire write.  It needs to be closed as soon as we get the columnar batch out that will be handed to the writer to close.

abellina replied:
Will fix

abellina replied:
Handled this and the one below in https://github.com/NVIDIA/spark-rapids/pull/8385/commits/e9765621c62e7aa2cd03ead6a47d258898be0fb0 and https://github.com/NVIDIA/spark-rapids/pull/8385/commits/48aec9468366c44386a45908f1090d14464141f6

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -939,68 +958,73 @@ class GpuDynamicPartitionDataConcurrentWriter(
+          }
+          // close the contiguous table
+          tabs(partIndex) = null
+          withResource(b) { _ =>
+            status.writerStatus.outputWriter.writeAndClose(cb, statsTrackers)
+          }
```
Holding `b` open but not referencing it in the block? Seems similar to other comment, need to close as soon as `cb` is built.

abellina replied:
yes, I got confused with ownership here: we took out the columns from the Table, and at that point it doesn't matter that we closed the ContiguousTable or the table_view.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -434,179 +464,164 @@ class GpuDynamicPartitionDataSingleWriter(
+        (cbKeys, outputColumnsTbl, partitionIndexes)
+      }
+    }
+    val splits = withResource(outputColumnsTbl) { _ =>
+      outputColumnsTbl.contiguousSplit(partitionIndexes: _*)
+    }
```
cbKeys are leaked if this throws.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -623,17 +638,22 @@ class GpuDynamicPartitionDataSingleWriter(
   }
 
+  private def writeBatchUsingCurrentWriterAndClose(batch: ColumnarBatch): Unit = {
+    statsTrackers.foreach(_.newBatch(currentWriterStatus.outputWriter.path(), batch))
```
If this throws then `batch` is leaked

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -165,43 +216,37 @@ class GpuSingleDirectoryDataWriter(
+            val partBatch = GpuColumnVector.from(partContigTable.getTable, dataTypes)
+            statsTrackers.foreach(_.newBatch(currentWriter.path(), partBatch))
+            recordsInFile += partBatch.numRows()
+            currentWriter.writeAndClose(partBatch, statsTrackers)
```
`writeAndClose` goes out of its way to close the batch early during processing, so any time we see a pattern like this:
```
withResource(x) {
  ...
  ...writeAndClose(..)
}
```
then something is probably wrong, because we're artificially holding onto something until the write of a batch completes, and a large part of the write is not on the GPU.  We need to close everything related to a batch except the batch itself, and then hand the batch to the writer so it can close it during the write processing at the appropriate time.

abellina commented:
Running all integration tests locally and will ping when all looks good. I tried addressing @jlowe's comments and refactored things a bit to deal with `ColumnarBatch` as that seemed a lot more direct in a lot of places.

abellina commented:
@jlowe @revans2 Note this is a bug fix with the changes I just fixed https://github.com/NVIDIA/spark-rapids/pull/8385/commits/65d0cdc4e665daa3f82581fe996730fc9a067bb7. I was using `zipWithIndex` instead of `zip` as the original code was, so I reverted to `zip` and added a comment.

Arguably I could have dropped a partition or a split point, so I can do that. But I am opting for less changes right now.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala:
```diff
@@ -434,179 +518,122 @@ class GpuDynamicPartitionDataSingleWriter(
-        val table = combined._2.getTable
-        val partPath = combined._1
-
+    // We have an entire batch that is sorted, so we need to split it up by key
+    // to get a batch per path
+    val (splits, paths) = splitBatchByKey(cb, partDataTypes)
+    withResource(splits) { _ =>
+      // NOTE: the `zip` here has the effect that will remove an extra `ContiguousTable`
+      // added at the end of `splits` because we use `upperBound` to find the split points,
+      // and the last split point is the number of rows.
+      splits.zip(paths).zipWithIndex.foreach { case ((partContigTable, partPath), partIx) =>
```
Seems like part of the confusion here is that splitBatchByKey should be returning table, path pairs rather than two separate arrays that this goes out of its way to put back together.

abellina replied:
Yeah, I'll zip them while in `splitBatchByKey` and move the comment there...

abellina commented:
build
